staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go

```diff
@@ -811,6 +811,22 @@ func reasonAndCodeForError(err error) (metav1.StatusReason, int32) {
 	return metav1.StatusReasonUnknown, 0
 }
 
+func CauseTypesForError(err error) ([]metav1.CauseType) {
+	var causeTypes []metav1.CauseType
+	if status := APIStatus(nil); errors.As(err, &status) {
+		var emptyStatusDetailsStruct *metav1.StatusDetails
+		if status.Status().Details != emptyStatusDetailsStruct{
+			details := status.Status().Details
+			statusCauses := details.Causes
+			for _, v := range statusCauses{
+				causeTypes = append(causeTypes, v.Type)
+			}
+			return causeTypes
+		}
+		return causeTypes
+	}
+	return causeTypes
+}
```

Comment:
@liggitt Could you please confirm that the code for this method is correct or not.
if yes,
- Are we thinking about adding a bool here in this condition
https://github.com/kubernetes/kubernetes/blob/cf9b5ab95a21e5b32b6aff72492a18ede37b785b/staging/src/k8s.io/kubectl/pkg/cmd/apply/patcher.go#L201-L203
with the existing bool like this 
```
if (err != nil && (errors.IsConflict(err) || errors.IsInvalid(err)) && p.Force) && newBool{ 
 	patchBytes, patchObject, err = p.deleteAndCreate(current, modified, namespace, name) 
} 
```
, where newBool return false if all the elements in causeTyes (returned from this helper method) are == `metav1.CauseTypeFieldChangeForbidden` and as a result will not execute `deleteAndCreate` function. 

Am I thinking correct here ?

Reply:
> where newBool return false if all the elements in causeTyes (returned from this helper method) are == metav1.CauseTypeFieldChangeForbidden and as a result will not execute deleteAndCreate function

exactly the opposite of that, actually... when talking to a server that knows how to return CauseTypeFieldChangeForbidden errors, we only want to deleteAndCreate if all the elements in the returned error have cause type CauseTypeFieldChangeForbidden

Deleting and recreating an object is destructive. we only want to do that if the creation of the new object will succeed, and the only errors that were previously returned were related to attempting to change field values in an update.